stdlib/LinearAlgebra/src/dense.jl

```diff
@@ -514,12 +514,38 @@ function (^)(A::AbstractMatrix{T}, p::Integer) where T<:Integer
 end
 function integerpow(A::AbstractMatrix{T}, p) where T
     TT = promote_op(^, T, typeof(p))
-    return (TT == T ? A : convert(AbstractMatrix{TT}, A))^Integer(p)
+    ATT = TT == T ? A : convert(AbstractMatrix{TT}, A)
+    return _integerpow(ATT, p)
 end
+_integerpow(A::AbstractMatrix, p) = A^Integer(p)
+function _integerpow(A::AbstractMatrix, p::Union{Float32, Float64})
```

Comment:
```suggestion
_integerpow(A::AbstractMatrix, p::Integer) = A^p
function _integerpow(A::AbstractMatrix, p)
```
It seems that the long definition here is the more general? Apart from `Float16` (although even there the overhead should be tiny), I'd feel safer with generic types avoiding the assumed-valid `Integer` conversion since that's how we got here in the first place.